### PR TITLE
[setup] autoconfigure libraries

### PR DESCRIPTION
**Pull request by xoviat**
Enables autodetection of library locations. Numpy can correctly detect various BLAS/lapack locations automatically rather than having to hardcode them in setup.py. In addition, we upgrade conda's outdated MKL library because the autoconfiguration prefers MKL if it's found and the outdated version lacked routines that SHTOOLS needed.

### CODE REVIEW

**Comment by MarkWieczorek** [comment]
I'm having a problem getting this to work on my system (OSX). I think it is related to this `get_info()`  statement for lapack which exits with an error in the console:
````
In [4]: lapack_info = get_info('lapack_opt')
ld: warning: ld: warning: ld: warning: ignoring file /usr/local/lib/libptcblas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libptcblas.aignoring file /usr/local/lib/libptf77blas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libptf77blas.aignoring file /usr/local/lib/libatlas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libatlas.a


Undefined symbols for architecture x86_64:
  "_ATL_buildinfo", referenced from:
      _main in _configtest.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ld: warning: ld: warning: ld: warning: ignoring file /usr/local/lib/libptcblas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libptcblas.aignoring file /usr/local/lib/libptf77blas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libptf77blas.aignoring file /usr/local/lib/libatlas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libatlas.a


Undefined symbols for architecture x86_64:
  "_ATL_buildinfo", referenced from:
      _main in _configtest.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
````
When installing locally with pip I get this output
````
    ignoring file /usr/local/lib/libatlas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libatlas.a

    Creating /usr/local/lib/python3.6/site-packages/pyshtools.egg-link (link to .)
    Adding pyshtools 4.1.dev0+2c3e0ad to easy-install.pth file

    Installed /Users/lunokhod/SphericalHarmonics/shtools-git
    ---- BUILDING DOCS ----
    creating pyshtools/doc
    Traceback (most recent call last):
      File "./pyshtools/make_docs.py", line 16, in <module>
        import _SHTOOLS
    ImportError: dlopen(/Users/lunokhod/SphericalHarmonics/shtools-git/pyshtools/_SHTOOLS.cpython-36m-darwin.so, 2): Symbol not found: _daxpy_
      Referenced from: /Users/lunokhod/SphericalHarmonics/shtools-git/pyshtools/_SHTOOLS.cpython-36m-darwin.so
      Expected in: flat namespace
     in /Users/lunokhod/SphericalHarmonics/shtools-git/pyshtools/_SHTOOLS.cpython-36m-darwin.so
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/Users/lunokhod/SphericalHarmonics/shtools-git/setup.py", line 289, in <module>
        setup(**metadata)
      File "/usr/local/lib/python3.6/site-packages/numpy/distutils/core.py", line 169, in setup
        return old_setup(**new_attr)
      File "/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python3.6/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python3.6/distutils/dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python3.6/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/Users/lunokhod/SphericalHarmonics/shtools-git/setup.py", line 146, in run
        check_call([sys.executable, doc_builder, doc_source, self.setup_path])
      File "/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python3.6/subprocess.py", line 291, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command '['/usr/local/opt/python3/bin/python3.6', './pyshtools/make_docs.py', '.', '.']' returned non-zero exit status 1.
  Rolling back uninstall of pyshtools
  Replacing /usr/local/lib/python3.6/site-packages/pyshtools.egg-link
  Rolling /usr/local/lib/python3.6/site-packages/easy-install.pth back to previous state
Cleaning up...
Command "/usr/local/opt/python3/bin/python3.6 -c "import setuptools, tokenize;__file__='/Users/lunokhod/SphericalHarmonics/shtools-git/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" develop --no-deps" failed with error code 1 in /Users/lunokhod/SphericalHarmonics/shtools-git/
Exception information:
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/local/lib/python3.6/site-packages/pip/commands/install.py", line 342, in run
    prefix=options.prefix_path,
  File "/usr/local/lib/python3.6/site-packages/pip/req/req_set.py", line 784, in install
    **kwargs
  File "/usr/local/lib/python3.6/site-packages/pip/req/req_install.py", line 845, in install
    install_options, global_options, prefix=prefix)
  File "/usr/local/lib/python3.6/site-packages/pip/req/req_install.py", line 1007, in install_editable
    show_stdout=False)
  File "/usr/local/lib/python3.6/site-packages/pip/utils/__init__.py", line 707, in call_subprocess
    % (command_desc, proc.returncode, cwd))
pip.exceptions.InstallationError: Command "/usr/local/opt/python3/bin/python3.6 -c "import setuptools, tokenize;__file__='/Users/lunokhod/SphericalHarmonics/shtools-git/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" develop --no-deps" failed with error code 1 in /Users/lunokhod/SphericalHarmonics/shtools-git/
````
I can probably just delete the offending file, but we should anticipate this to happen on other systems as well...

**Comment by MarkWieczorek** [comment]
I just confirmed that deleting the old libatlas file works.

**Comment by xoviat** [comment]
Can you build scipy with the corrupted ATLAS?

**Comment by MarkWieczorek** [comment]
I don't know, I think that the scipy pip installer is for a wheel, so there were no errors. I could try downloading the scipy code and installing it manually, but this will have to wait until tuesday.

**Comment by xoviat** [comment]
For building scipy, it's:

```
pip install git+https://github.com/scipy/scipy.git
```

**Comment by MarkWieczorek** [comment]
That pip install command gave all kinds of unrelated errors and then crashed, so I downloaded the code and installed it myself. Everything appears to have worked, but I do get a lot of warnings about the libatlas file, such as

````
ld: warning: ignoring file /usr/local/lib/libf77blas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libf77blas.a
ld: warning: ignoring file /usr/local/lib/libcblas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libcblas.a
ld: warning: ignoring file /usr/local/lib/libatlas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libatlas.a
ld: warning: ld: warning: ld: warning: ignoring file /usr/local/lib/libf77blas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libf77blas.aignoring file /usr/local/lib/libcblas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libcblas.aignoring file /usr/local/lib/libatlas.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libatlas.a
````
At the end of installation, i get 
````
     Installed /Users/lunokhod/Desktop/scipy-1.0.0
    /usr/local/lib/python3.6/site-packages/numpy/distutils/system_info.py:1105: UserWarning:
    *********************************************************************
        Could not find lapack library within the ATLAS installation.
    *********************************************************************

      warnings.warn(message)
    Running from scipy source directory.
    /usr/local/lib/python3.6/site-packages/numpy/distutils/system_info.py:639: UserWarning: Specified path /usr/local/include/python3.6m is invalid.
      warnings.warn('Specified path %s is invalid.' % d)
    /usr/local/lib/python3.6/site-packages/numpy/distutils/system_info.py:639: UserWarning: Specified path /opt/local/include/python3.6m is invalid.
      warnings.warn('Specified path %s is invalid.' % d)
    "object of type 'type' has no len()" in evaluating 'len(list)' (available names: [])
    "object of type 'type' has no len()" in evaluating 'len(list)' (available names: [])
    "object of type 'type' has no len()" in evaluating 'len(list)' (available names: [])
    "object of type 'type' has no len()" in evaluating 'len(list)' (available names: [])
    "object of type 'type' has no len()" in evaluating 'len(list)' (available names: [])
    "object of type 'type' has no len()" in evaluating 'len(list)' (available names: [])
Successfully installed scipy
Cleaning up...
````
So, what do you suggest? I delete the offending libraries and forget about it? Or should we try to solve this problem? Maybe we just need better error handling?

**Comment by xoviat** [comment]
First of all, the pip command should have worked on OSX (but not sure it's worth spending time of resolving that). Second all, the reason that I asked to check building SciPy is that it also uses this function to configure itself:

https://github.com/scipy/scipy/blob/749e7152e2a29f7d806acc3407c791fd44e1de95/scipy/optimize/_trlib/setup.py#L10

Honestly there is not enough debugging information here to go on. NumPy should have identified that the ATLAS library was bad and ignored it, and I don't understand why it would work for SciPy and not for SHTOOLS (it's the same function, after all). It could be a bug in NumPy, but I don't think it's an efficient use of anyone's time to investigate why this happened though.

> Maybe we just need better error handling?

The idea is that you can just call that function and not have to worry about errors. So if it were a problem, the errors would be handled in NumPy.

In summary, I think it's better to delete the file and continue, unless you want to spend time debugging. Here's the relevant function in case you want to attach a debugger and figure out why it's not ignoring ATLAS when it should:

https://github.com/numpy/numpy/blob/75b2d5d427afdb1392f2a0b2092e0767e4bab53d/numpy/distutils/system_info.py#L1501
